### PR TITLE
Make logging an optional feature and disable it by default

### PR DESCRIPTION
**Pull request by akosthekiss**
JerryScript-DCO-1.0-Signed-off-by: Akos Kiss akiss@inf.u-szeged.hu

### CODE REVIEW

**Comment by akosthekiss** [comment]
Results on binary size (and on RSS in consequence):

Benchmark | RSS (bytes) |
----: | ----: | 
3d-cube.js | 53248 -> 53248 : 0.000% | 
3d-raytrace.js | 106496 -> 106496 : 0.000% | 
access-binary-trees.js | 49152 -> 49152 : 0.000% | 
access-fannkuch.js | 20480 -> 16384 : +20.000% | 
access-nbody.js | 24576 -> 24576 : 0.000% | 
bitops-3bit-bits-in-byte.js | 16384 -> 12288 : +25.000% | 
bitops-bits-in-byte.js | 16384 -> 12288 : +25.000% | 
bitops-bitwise-and.js | 12288 -> 12288 : 0.000% | 
bitops-nsieve-bits.js | 110592 -> 110592 : 0.000% | 
controlflow-recursive.js | 86016 -> 81920 : +4.762% | 
crypto-aes.js | 65536 -> 61440 : +6.250% | 
crypto-md5.js | 131072 -> 126976 : +3.125% | 
crypto-sha1.js | 81920 -> 81920 : 0.000% | 
date-format-tofte.js | 36864 -> 36864 : 0.000% | 
date-format-xparb.js | 40960 -> 36864 : +10.000% | 
math-cordic.js | 16384 -> 12288 : +25.000% | 
math-partial-sums.js | 16384 -> 12288 : +25.000% | 
math-spectral-norm.js | 20480 -> 20480 : 0.000% | 
string-base64.js | 139264 -> 139264 : 0.000% | 
string-fasta.js | 28672 -> 28672 : 0.000% | 
Geometric mean: | +7.801% | 

Binary sizes (bytes)
1ac2903d3c:132784
d60947a9c4:130908

(~1.8k gain)

File: jerry-core/CMakeLists.txt [hunk 1]
```diff
@@ -66,6 +67,12 @@ if(JERRY_CMDLINE_SNAPSHOT)
   set(FEATURE_SNAPSHOT_SAVE_MESSAGE " (FORCED BY SNAPSHOT TOOL)")
 endif()
 
+if(FEATURE_DEBUGGER OR FEATURE_MEM_STATS OR FEATURE_PARSER_DUMP OR FEATURE_REGEXP_DUMP)
+  set(FEATURE_LOGGING ON)
```
Comment (zherczeg) [inline jerry-core/CMakeLists.txt]: Debugging should not force logging.

File: jerry-core/parser/js/common.c [hunk 2]
```diff
@@ -74,6 +76,7 @@ util_print_number (ecma_number_t num_p) /**< number to print */
   lit_utf8_byte_t str_buf[ECMA_MAX_CHARS_IN_STRINGIFIED_NUMBER];
   lit_utf8_size_t str_size = ecma_number_to_utf8_string (num_p, str_buf, sizeof (str_buf));
   str_buf[str_size] = 0;
+  JERRY_UNUSED (str_buf); // potentially
```
Comment (zherczeg) [inline jerry-core/parser/js/common.c]: We will forget adding these macros. Just like asserts, logging should reference its arguments.

File: jerry-core/parser/js/js-parser.c [hunk 3]
```diff
@@ -2637,9 +2642,9 @@ parser_parse_function (parser_context_t *context_p, /**< context */
   if (context_p->is_show_opcodes)
   {
 #ifndef CONFIG_DISABLE_ES2015_CLASS
-    bool is_constructor = context_p->status_flags & PARSER_CLASS_CONSTRUCTOR;
-    JERRY_DEBUG_MSG (is_constructor ? "\n--- Class constructor parsing end ---\n\n"
-                                    : "\n--- Function parsing end ---\n\n");
+    JERRY_DEBUG_MSG ("\n--- %s parsing end ---\n\n",
+                     context_p->status_flags & PARSER_CLASS_CONSTRUCTOR ? "Class constructor"
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser.c]: Brackets around `context_p->status_flags & PARSER_CLASS_CONSTRUCTOR`.

**Comment by akosthekiss** [comment]
PR updated according to the reviews

Note 1: "Just like asserts, logging should reference its arguments." This needed some macro black magic as the classic `JERRY_UNUSED` doesn't work right with variadic arguments.

Note 2: As of late, the ESP target has been erroring. It makes all PRs red (including this one). Hope it's only some instability that goes away soon.

**Comment by akosthekiss** [comment]
Commits squashed.

**Pull request merged**